python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py

```diff
@@ -364,3 +414,291 @@ def _optimizer_step(self):
+        virtual_pp_rank = self._get_virtual_pp_rank(micro_step, forward=True)
+        self.set_virtual_pipeline_rank(virtual_pp_rank)
+
+        # some checkers
+        assert hasattr(self, 'input_tensors')
+        assert hasattr(self, 'output_tensors')
+        if not self._forward_only:
+            assert hasattr(self, 'output_tensor_grads')
+
+        if self.is_pipeline_first_stage():
+            if len(self.input_tensors[virtual_pp_rank]) == len(
+                    self.output_tensors[virtual_pp_rank]):
+                self.input_tensors[virtual_pp_rank].append(None)
+        input_tensor = self.input_tensors[virtual_pp_rank][-1]
+        output_tensor = self._forward_step(input_tensor, virtual_pp_rank)
+        self.output_tensors[virtual_pp_rank].append(output_tensor)
+
+        if self._forward_only:
+            # no need to store tensor for backward
+            self.input_tensors[virtual_pp_rank].pop()
+            self.output_tensors[virtual_pp_rank].pop()
+
+        return output_tensor
+
+    def _backward_step_helper(self, micro_step):
+        virtual_pp_rank = self._get_virtual_pp_rank(micro_step, forward=False)
+        self.set_virtual_pipeline_rank(virtual_pp_rank)
+
+        # some checkers
+        assert hasattr(self, 'input_tensors')
+        assert hasattr(self, 'output_tensors')
+        assert hasattr(self, 'output_tensor_grads')
+
+        if self.is_pipeline_last_stage():
+            if len(self.output_tensor_grads[virtual_pp_rank]) == 0:
+                self.output_tensor_grads[virtual_pp_rank].append(None)
+
+        input_tensor = self.input_tensors[virtual_pp_rank].pop(0)
+        output_tensor = self.output_tensors[virtual_pp_rank].pop(0)
+        output_tensor_grad = self.output_tensor_grads[virtual_pp_rank].pop(0)
+        input_tensor_grad = self._backward_step(input_tensor, output_tensor,
+                                                output_tensor_grad)
+
+        return input_tensor_grad
+
+    def interleave_pipeline(self,
+                            data,
+                            scaler,
+                            forward_only=False,
+                            compute_loss=True):
+        # use interleave scheduling strategy.
+        # this strategy is inspired by:
+        # https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/schedules.py
+        if not compute_loss:
+            assert not forward_only, "compute_loss can only be set to False when forward_only is set to True"
+
+        # init some attributes for this batch run
+        self.scaler = scaler
+        self.data = data
+        self.total_loss = None
+        self.micro_batch_id = 0
+        self._forward_only = forward_only
+
+        # init some data buffers for interleave scheduler
+        self.input_tensors = [[] for _ in range(self.num_model_chunks)]
+        self.output_tensors = [[] for _ in range(self.num_model_chunks)]
+        self.output_tensor_grads = [[] for _ in range(self.num_model_chunks)]
```

Comment:
Not quite important: `self.output_tensor_grads` is not necessary when `forward_only = True`.